kvstoremesh/metrics/metrics.go

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package metrics
+
+import (
+	"errors"
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/collectors"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/spf13/pflag"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/metrics/metric"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "metrics")
+)
+
+var Cell = cell.Module(
+	"kvstoremesh-metrics",
+	"KVStoreMesh metrics",
+
+	cell.Config(MetricsConfig{}),
+	cell.Invoke(registerMetricsManager),
+)
+
+type MetricsConfig struct {
+	// PrometheusServeAddr IP:Port on which to serve prometheus metrics (pass ":Port" to bind on all interfaces, "" is off)
+	PrometheusServeAddr string
+}
+
+func (def MetricsConfig) Flags(flags *pflag.FlagSet) {
+	flags.String(option.PrometheusServeAddr, def.PrometheusServeAddr, "Address to serve Prometheus metrics")
+}
+
+type metricsManager struct {
+	registry *prometheus.Registry
+	server   http.Server
+
+	metrics []metric.WithMetadata
+}
+
+type params struct {
+	cell.In
+
+	MetricsConfig
+
+	Metrics []metric.WithMetadata `group:"hive-metrics"`
+}
+
+func registerMetricsManager(lc hive.Lifecycle, params params) error {
+	manager := metricsManager{
+		registry: prometheus.NewPedanticRegistry(),
+		server:   http.Server{Addr: params.PrometheusServeAddr},
+		metrics:  params.Metrics,
+	}
+
+	if params.PrometheusServeAddr != "" {
+		lc.Append(&manager)
+	} else {
+		log.Info("Prometheus metrics are disabled")
+	}
+
+	return nil
+}
+
+func (mm *metricsManager) Start(hive.HookContext) error {
+	log.Info("Registering metrics")
+
+	mm.registry.MustRegister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}))
+	mm.registry.MustRegister(collectors.NewGoCollector())
```

Comment:
In the Agent we recently enabled sched latency metrics: https://github.com/cilium/cilium/blob/main/pkg/metrics/registry.go#L116 by default.

The idea was that when people run Cilium in severely CPU constrained environments that we sometimes begin to see strange behaviour as goroutines get large scheduling latency.

Might be worth something to have here too, keep things consistent and we can easily build alerting around that.

Reply:
Added, thanks for the heads up.